sql/core/src/main/scala/org/apache/spark/sql/execution/script/ScriptTransformationExec.scala

```diff
@@ -0,0 +1,313 @@
+
+    child.execute().mapPartitions { iter =>
+      if (iter.hasNext) {
+        val proj = UnsafeProjection.create(schema)
+        processIterator(iter, broadcastedHadoopConf.value).map(proj)
+      } else {
+        // If the input iterator has no rows then do not launch the external script.
+        Iterator.empty
+      }
+    }
+  }
+
+  def checkFailureAndPropagate(writerException: Option[Throwable],
+                               cause: Throwable = null,
+                               proc: Process,
+                               stderrBuffer: CircularBuffer): Unit = {
+    if (writerException.isDefined) {
+      throw writerException.get
+    }
+
+    // Checks if the proc is still alive (incase the command ran was bad)
+    // The ideal way to do this is to use Java 8's Process#isAlive()
+    // but it cannot be used because Spark still supports Java 7.
+    // Following is a workaround used to check if a process is alive in Java 7
+    // TODO: Once builds are switched to Java 8, this can be changed
+    try {
+      val exitCode = proc.exitValue()
+      if (exitCode != 0) {
+        logError(stderrBuffer.toString) // log the stderr circular buffer
+        throw new SparkException(s"Subprocess exited with status $exitCode. " +
+          s"Error: ${stderrBuffer.toString}", cause)
+      }
+    } catch {
+      case _: IllegalThreadStateException =>
+      // This means that the process is still alive. Move ahead
+    }
+  }
+
+  def createReader(inputStream: InputStream): BufferedReader =
+    new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8))
+
+  def processIterator(
+      inputIterator: Iterator[InternalRow],
+      hadoopConf: Configuration) : Iterator[InternalRow]
+}
+
+private[sql] class ScriptTransformationWriterThread(
+    iter: Iterator[InternalRow],
+    inputSchema: Seq[DataType],
+    outputProjection: Projection,
+    ioschema: ScriptTransformIOSchema,
+    outputStream: OutputStream,
+    proc: Process,
+    stderrBuffer: CircularBuffer,
+    taskContext: TaskContext,
+    conf: Configuration
+  ) extends Thread("Thread-ScriptTransformation-Feed") with Logging with Serializable {
+
+  setDaemon(true)
+
+  @volatile protected var _exception: Throwable = null
+
+  /** Contains the exception thrown while writing the parent iterator to the external process. */
+  def exception: Option[Throwable] = Option(_exception)
+
+  protected def init(): Unit = {
+    TaskContext.setTaskContext(taskContext)
+  }
+
+  protected def processRow(row: InternalRow, numColumns: Int): Unit = {
+    val data = if (numColumns == 0) {
+      ioschema.inputRowFormatMap("TOK_TABLEROWFORMATLINES")
+    } else {
+      val sb = new StringBuilder
+      sb.append(row.get(0, inputSchema(0)))
+      var i = 1
+      while (i < numColumns) {
+        sb.append(ioschema.inputRowFormatMap("TOK_TABLEROWFORMATFIELD"))
+        sb.append(row.get(i, inputSchema(i)))
```

Comment:
@hvanhovell : for now I am not going the code-gen route as it will make the diff bigger and delay things. I am doing what PR #14279 did and adding code-gen as `Future TODOs` in PR description.